sources/segnet_keras.py

```diff
@@ -0,0 +1,216 @@
+    seed=SEED)
+validation_images_generator = train_datagen.flow_from_directory(
+    VALIDATION,
+    classes=['images'],
+    target_size=(SIZE, SIZE),
+    batch_size=BATCH_SIZE,
+    class_mode=None,
+    seed=SEED)
+validation_mask_generator = train_datagen.flow_from_directory(
+    VALIDATION,
+    classes=['labels'],
+    target_size=(SIZE, SIZE),
+    batch_size=BATCH_SIZE,
+    color_mode='grayscale',
+    class_mode=None,
+    seed=SEED)
+masks_generator = iter(to_categorical(x, label_ids) for x in masks_generator)
+train_generator = zip(images_generator, masks_generator)
+validation_mask_generator = iter(to_categorical(x, label_ids)
+                                 for x in validation_mask_generator)
+validation_generator = zip(validation_images_generator, validation_mask_generator)
+
+# Input layer. Squarred RGB image.
+img_input = Input(shape=(SIZE, SIZE, 3))
+
+# encoder part
+x = block_encoder(img_input, 32)
+x = block_encoder(x, 64)
+x = block_encoder(x, 128)
+x = block_encoder(x, 256)
+
+# decoder part
+x = block_decoder(x, 256)
+x = block_decoder(x, 128)
+x = block_decoder(x, 64)
+x = block_decoder(x, 32)
+# Number of feature maps equals to the number of classes, i.e. labels
+x = Conv2DTranspose(CLASSES, kernel_size=2, padding='same')(x)
+
+# Last layer
+y = Activation('softmax', name='target')(x)
+
+model = Model(img_input, y)
+model.compile(loss='categorical_crossentropy',
+              optimizer='adam',
+              metrics=['acc', 'mae'])
+json_model = model.to_json()
+with open(osp.join(OUTPUT_DIR, "segnet-model-keras.json"), "w") as fobj:
+    fobj.write(json_model)
+print(model.summary())
+print("See output dir {}".format(OUTPUT_DIR))
+tsboard = keras.callbacks.TensorBoard(log_dir=osp.join(OUTPUT_DIR, 'log'), write_graph=True)
+check = keras.callbacks.ModelCheckpoint(
+    osp.join(OUTPUT_DIR, "checkpoint-segnet-epoch-{epoch:02d}-valacc-{val_acc:.2f}.h5"),
+    monitor='val_acc',
+    verbose=0,
+    save_best_only=True,
+    save_weights_only=False,
+    mode='auto', period=1)
+earlystop = keras.callbacks.EarlyStopping(monitor='val_acc',
+                                          min_delta=0.002,
+                                          patience=10,
+                                          verbose=0,
+                                          mode='auto')
+hist = model.fit_generator(train_generator, epochs=EPOCHS,
+                           callbacks=[tsboard, check, earlystop],
+                           steps_per_epoch=STEPS,
+                           validation_data=validation_generator,
+                           validation_steps=VALIDATION_STEPS)
+metrics = {"epoch": hist.epoch,
+           "metrics": hist.history,
+           "params": hist.params}
+with open(osp.join(OUTPUT_DIR, "hist-metrics.json"), "w") as fobj:
+    json.dump(metrics, fobj)
+fname = "final-segnet-batch-{}-size-{}-max-epoch-{}.h5"
+fname = fname.format(BATCH_SIZE, SIZE, EPOCHS)
+model.save(osp.join(OUTPUT_DIR, fname))
```

Comment:
Same remark as before regarding output backup. Moreover what's the difference between this `h5` file and checkpoint files? Same nature, but different training state ? Or the final segnet result includes more features ?

Reply:
Yes, it's the same nature. Just the final one :)